### ISSUE TITLE
Reduce the extension size by accessing music resources remotely

### ISSUE THREAD

**Issue by jordigarcl** [PikaDude/Animal-Crossing-Music-Extension#5]
This extension has a very big download size because it relies on storing music resources on every user's local storage.

These music resources can easily account for more than 95% of the current extension size.

I believe we should evaluate the possibility of dynamically retrieving these resources from some remote source. Then, we could provide an option to store them locally when retrieved. That way, we could still provide offline support while having a very small extension size!

I am aware there are certain cons and setbacks for adopting this change (see [this discussion](https://github.com/animal-crossing-music-extension/Animal-Crossing-Music-Extension/issues/13) from the original project), but the extension size is already giving problems to users when it comes to download it from the Chrome store, and future features will probably increase the number of music assets.

**Comment by jordigarcl** [comment]
This is a big change that must be analysed before integrating it in the project.

I'm going to work on a 'proof of concept' for this feature, to review how it could be implemented 👍

**Comment by jordigarcl** [comment]
I did some research on this.

**TL;DR:** There are more drawbacks that I anticipated and I don't know anymore if this is something that _must_ be done. I don't believe that switching to remote audio is worth the hassle.

### Current concerns
#### Large extension size
Yes, the extension size is big, and this translates into a big initial download. This however effectively provides full offline support. 

Given the drawbacks found on the evaluated alternatives, this may the best we can do, and is not that big of a deal.

#### Music decoupled from public repo
We can offer the audio sources separately and give clear instructions on how to include them in the project. 

I don't think this would be a lot task for. If this the direction we are going, we should consider being explicit about it in the code. What about displaying an informative message if the audio sources are not detected?

#### Webstore has a 2GB size limit 
[Developer Chrome FAQ - What is the size limit for apps in the store?](https://developer.chrome.com/webstore/faq)

Currently apps published in the Chrome Webstore have a 2GB file size limit.

The current `develop` version (8a7aeb4), once packed into a `.crx` file, weights around 285MB. 

Even though it is still missing `kk`, `events` and `new-horizons` music, I think this is a limitation we can comfortable work with.

### Problems going remote-first
#### Playing remote audio
As long as the user has a decent internet connection, it is feasible to play remote audio files with seamless looping. This was not a concern here.

#### Download songs silently:
I toyed with the idea of downloading each song to the local storage the first time it was played, but then the user would only have the hourly songs he/she had already listened to. 

We could trigger a batch download when the user first installs the extension, but we would be back to square one (except for the fact that they would be decoupled from the code) and we wouldn't be leveraging the Chrome Webstore 'free' hosting.

And a **BIG** point here is that Chrome [doesn't support "silent" downloads](https://stackoverflow.com/questions/36007018/downloading-silently-by-chrome-extension). If the extension were to download a remote audio file, the user would see it. What is more, we could [only download a file](https://stackoverflow.com/questions/33836049/how-do-i-download-a-file-to-the-extension-folder-in-a-chrome-extension) under the user-defined Download directory, which the extension doesn't have access to.

#### Native host app
Chrome allows communication between [extensions and native hosts](https://developer.chrome.com/apps/nativeMessaging), that is, native OS applications. Theoretically, an extension would be able to make the native app download files to the user local storage and query these local files by means of passing messages.

Drawbacks here: 
- The native host app [can only send 1MB at a time](https://stackoverflow.com/questions/27243529/native-messaging-host-not-able-to-send-1-mb-data) to Chrome. Most of the `.ogg` files are >1MB so this would require an extra effort of splitting the messages in chunks before sending them and then putting them back together when receiving them.
- This native host must be [distributed separately](https://stackoverflow.com/questions/45568151/how-to-distribute-a-native-messaging-host-with-chrome-extension). Users would have to download and install an executable outside of the Chrome Webstore.
- The native app must, well, be native. Though it is a very simple use case, this means another codebase to maintain for each OS we want to support.
- These messages work over JSON. We would need to encode and decode the audio files with base64

<br>
So, all in all, I would stay with the current approach 🤷
If we ever are forced to remove the audio files from the packaged extension, playing remote audio is fine.

**Comment by PikaDude** [comment]
I agree. The main decision is would we prefer either offline listening or a smaller extension size?

Most people use Chrome as a web browser, which typically means they're connected to the Internet when using it, so I guess we would prioritise a smaller extension size.

However, due to the limitations of what Chrome Extensions can do, downloading and saving songs for offline use is not possible, so our only option is to just store them on a remote server and download them temporarily as-need-be (simply by setting the audio src URL to a remote server).

The extension would only really need to fetch the song every time the song changes (if we set the remote server cache settings to a fairly high number, hopefully causing the client to cache the songs, it could potentially be even less) meaning that this solution is still a viable option. This will overall reduce the extension size by a tonne, at the cost of requiring an Internet connection when listening. 


From memory, if the audio file is cached due to how browsers like Chrome cache files to save on bandwidth, it could potentially still work offline if the user has listened to the song before and the cache has not expired, I'll give this a shot and see how it goes.

**Comment by PikaDude** [comment]
I gave this a shot in f9852ff2d8ca4360bbb7288d01571ef2fcd26224. It all seems to work okay. I'll copy my commit description here for easy reference. We could use this as a solution.

```
Songs are now stored remotely on my server, and
are cached locally on the client's computer
for 5 days. They will continue to work offline even
if the song changes, or the user restarts chrome.
The only catch is that the user must listen to the
song at least once for it to be cached, otherwise
it will not play. Additionally, the songs must be
renewed every 5 days, so the offline cache is not
permanent. 5 days was just a test duration, I can
extend it if we deem it necessary.
```

**Comment by jordigarcl** [comment]
Yeah, accessing remote audio files seem to work fine without any changes.

My concern is whether your server will be able to handle the load of 45K potential users (though of course in reality it will be way less than that). 

I did not know Chrome was able to cache this kind of songs. In this case most concerns are gone! How can we control this cache TTL? 

> I agree. The main decision is would we prefer either offline listening or a smaller extension size?

I **personally** think providing offline support is a core feature of the extension. Working offline means you don't have access to online music/youtube so I think being able to keep listening to music provides a lot of value here.

As I said, a bigger extension size provides a lot of benefits (offline, robust agains takedowns) and not that many drawbacks (>1GB of storage, a larger initial download).

**That being said**, if we choose to go to this direction I'll support it 👍 and we can always find a way to provide extraofficial offline support to power users.

**BTW** thank you for offering your server to try this! I was about to set up a raspi but seeing how it would not escalate I end up setting an ephemeral local web server.

**Comment by PikaDude** [comment]
My server runs behind Cloudflare, which will cache the resources and save on bandwidth. It's my own VPS that I use to run various projects, and I don't mind hosting everything on there.

Web browsers are generally set to cache resources if they are told to. This saves bandwidth for the user and the server, and also decreases loading times. We can control this cache TLL by adjusting the "cache-control" header of the HTTP response. Right now it's set to expire after 5 days, although that can easily be adjusted.

We have to look into the target demographic here. 100% of users use the extension in the web browser Google Chrome. Chrome is rendered effectively useless when not connected to the Internet, so users will not use it that often when not connected. Secondly, this extension is made for PCs. The only time a user would really be offline is when they're out and about on their laptop, and I don't believe a lot of people would really use it that often while they're out and about on their laptop. However, a larger majority of users are wary of storage space. While caching songs will use space, it will definitely use a lot less as it will only cache songs that the user listens to. Users will no longer need to necessarily download and also store all of K.K.'s songs, and songs from the other games if they don't ever listen to them. Also some users reported issues trying to download such a large extension (although that may just have been an issue on their end, who knows).

Regardless, this solution presents partial offline support, as songs that the user has listened to will be cached and continue to work offline - meaning that total offline support will not be lost.

Being robust against takedowns isn't really a valid argument as nothing is exactly stopping the entire extension being ripped down from the Chrome Webstore, or the Google Drive link also being shot down (these have also happened to others in the past, and are not at all protected).

Lastly, 5 days was just a test. These files aren't planned to change or anything, so we can crank it to anything we want, 14 days, 30 days, a year or even infinite (I wouldn't recommend the last two, personally).

**Comment by paladique** [comment]
>  It's my own VPS that I use to run various projects, and I don't mind hosting everything on there.

Thanks for offering! Would love to have a little donation fund set up (maybe linking to your Ko-fi?) once we merge this and release to support your hosting!

**Comment by PikaDude** [comment]
Nah, it's fine. It's really cheap, and I'm happy to pay for it myself. Plus, I don't really want to go in that direction where we ask for money or anything like that.

**Comment by PikaDude** [comment]
Also, because of this solution, we can increase the quality of the music again. I'll go through and do them all again.

**Comment by jordigarcl** [comment]
I agree to going remote-first 👍
